src/onyx/peer/function.clj

```diff
@@ -66,21 +62,66 @@
                 (count active-peers)))
       (rand-nth (operation/select-n-peers id active-peers max-downstream-links)))))
 
-;; Needs a performance boost
-(defmethod p-ext/write-batch :default
-  [{:keys [onyx.core/id onyx.core/results onyx.core/messenger onyx.core/job-id onyx.core/max-downstream-links] :as event}]
-  (let [leaves (fast-concat (map :leaves results))
-        egress-tasks (:egress-ids (:onyx.core/serialized-task event))]
-    (when-not (empty? leaves)
-      (let [replica @(:onyx.core/replica event)
-            segments (build-segments-to-send leaves)
-            groups (group-by (juxt :route :hash-group) segments)
-            allocations (get (:allocations replica) job-id)]
-        (doseq [[[route hash-group] segs] groups]
-          (let [peers (get allocations (get egress-tasks route))
-                active-peers (filter #(= (get-in replica [:peer-state %]) :active) peers)
-                target (pick-peer id active-peers hash-group max-downstream-links)]
-            (when target
-              (let [link (operation/peer-link event target)]
-                (onyx.extensions/send-messages messenger event link segs)))))
-        {}))))
+(defn read-batch [{:keys [onyx.core/messenger] :as event}]
```

Comment:
Hmm yes, though I think I will want to export it through interop so that it can be used by output plugins in java (see what I mean in write-batch below)